### PR TITLE
Document the new NodeInclusionPolicyInPodTopologySpread feature gate

### PR DESCRIPTION
**Pull request by kerthcet**
Signed-off-by: kerthcet <kerthcet@gmail.com>

Ref: https://github.com/kubernetes/enhancements/issues/3094
/sig scheduling

### CODE REVIEW

**Review by chrisnegus — COMMENTED** [review]
/lgtm

**Comment by sftim** [comment]
This doesn't seem to be ready for review.
/lgtm cancel

**Comment by kerthcet** [comment]
Document is ready to review now.  cc @alculquicondor

**Comment by cathchu** [comment]
Hey there 👋 @kerthcet, 1.25 Release Docs Shadow here!

Thanks for filing this placeholder. As a reminder, final docs PRs are due August 9th. Take a look at [Documenting a release](https://kubernetes.io/docs/contribute/new-content/new-features/#for-developers-or-other-sig-members) for additional information for the docs requirement for the release.

Thanks again!

**Comment by kerthcet** [comment]
Resolve the conflicts caused by https://github.com/kubernetes/website/pull/33792

**Comment by kerthcet** [comment]
cc @sftim PTAL

File: content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md [hunk 1]
```diff
@@ -64,6 +64,8 @@ spec:
       topologyKey: <string>
       whenUnsatisfiable: <string>
       labelSelector: <object>
+      nodeAffinityPolicy: <string> # optional; alpha since v1.25
```
Comment (alculquicondor) [inline content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md:67]: it's not an arbitrary string. I would say something like `[Honor|Ignore]`

Also, in line 51, it says `Here is an example`. But it's not an example. Maybe it should say, something like: `The usage of this field looks like the following:`

File: content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md [hunk 2]
```diff
@@ -123,6 +125,33 @@ your cluster. Those fields are:
   See [Label Selectors](/docs/concepts/overview/working-with-objects/labels/#label-selectors)
   for more details.
 
+- **nodeAffinityPolicy** indicates how we will treat Pod's nodeAffinity/nodeSelector
+when calculating pod topology spread skew. Options are:
+  - Honor: only nodes matching nodeAffinity/nodeSelector are included in the calculations.
+  - Ignore: nodeAffinity/nodeSelector are ignored. All nodes are included in the calculations.
+
+  If this value is nil, the behavior is equivalent to the Honor policy.
```
Comment (alculquicondor) [inline content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md:133]:
```suggestion
  If this value is null, the behavior is equivalent to the Honor policy.
```

In documentation, we use the values as if it was written in json/yaml.

File: content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md [hunk 3]
```diff
@@ -123,6 +125,33 @@ your cluster. Those fields are:
...
+  in order to use it.
+  {{< /note >}}
+
+- **nodeTaintsPolicy** indicates how we will treat node taints when calculating
+pod topology spread skew. Options are:
+  - Honor: nodes without taints, along with tainted nodes for which the incoming pod
+	  has a toleration, are included.
+  - Ignore: node taints are ignored. All nodes are included.
+
+  If this value is nil, the behavior is equivalent to the Ignore policy.
```
Comment (alculquicondor) [inline content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md:147]: null

File: content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md [hunk 4]
```diff
@@ -123,6 +125,33 @@ your cluster. Those fields are:
   See [Label Selectors](/docs/concepts/overview/working-with-objects/labels/#label-selectors)
   for more details.
```
Comment (alculquicondor) [inline content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md:126]: We have improved our API docs a lot to include definitions of "topology domain", "eligible domain", and others. For example: https://github.com/kubernetes/api/blob/c8f06018bfc8e3c8213c53ec87b7fa4e2b9121ab/core/v1/types.go#L3396

Could you add those in this doc?

Maybe we could add the definitions before talking about the field themselves.

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 5]
```diff
@@ -156,6 +156,7 @@ different Kubernetes components.
 | `NetworkPolicyEndPort` | `false` | Alpha | 1.21 | 1.21 |
 | `NetworkPolicyEndPort` | `true` | Beta | 1.22 |  |
 | `NetworkPolicyStatus` | `false` | Alpha | 1.24 |  |
+| `NodeInclusionPolicyInPodTopologySpread` | `false` | Alpha | 1.25 |
```
Comment (alculquicondor) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:159]:
```suggestion
| `NodeInclusionPolicyInPodTopologySpread` | `false` | Alpha | 1.25 | |
```

File: content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md [hunk 6]
```diff
@@ -556,7 +585,7 @@ section of the enhancement proposal about Pod topology spread constraints.
   cluster. This could lead to a problem in autoscaled clusters, when a node pool (or
   node group) is scaled to zero nodes, and you're expecting the cluster to scale up,
   because, in this case, those topology domains won't be considered until there is
-  at least one node in them.
```
Comment (sftim) [inline content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md:559]: Please retain the line break here (triggered by two spaces at the end of the line).

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 7]
```diff
@@ -1008,6 +1009,9 @@ Each feature gate is designed for enabling/disabling a specific feature:
 - `NetworkPolicyEndPort`: Enable use of the field `endPort` in NetworkPolicy objects,
   allowing the selection of a port range instead of a single port.
 - `NetworkPolicyStatus`: Enable the `status` subresource for NetworkPolicy objects.
+- `NodeInclusionPolicyInPodTopologySpread`: Enable using `nodeAffinityPolicy` and `nodeTaintsPolicy`
+  in Pod [topology spread constraints](/docs/concepts/workloads/pods/pod-topology-spread-constraints/)
```
Comment (sftim) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:1013]:
```suggestion
  in [Pod topology spread constraints](/docs/concepts/scheduling-eviction/topology-spread-constraints/)
```

File: content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md (hunk 4 above)
Comment (sftim) [reply content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md:126]: It's also OK to link to https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#scheduling (ie, in Markdown it'd be `[anchor text](/docs/reference/kubernetes-api/workload-resources/pod-v1/#scheduling)`) if that's useful. No need to repeat the API reference word-for-word.

File: content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md (hunk 2 above)
Comment (kerthcet) [reply content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md:133]: Yeah.

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 8]
```diff
@@ -1010,7 +1010,7 @@ Each feature gate is designed for enabling/disabling a specific feature:
   allowing the selection of a port range instead of a single port.
 - `NetworkPolicyStatus`: Enable the `status` subresource for NetworkPolicy objects.
 - `NodeInclusionPolicyInPodTopologySpread`: Enable using `nodeAffinityPolicy` and `nodeTaintsPolicy`
-  in Pod [topology spread constraints](/docs/concepts/workloads/pods/pod-topology-spread-constraints/)
+  in Pod [topology spread constraints](/docs/concepts/scheduling-eviction/topology-spread-constraints/)
```
Comment (sftim) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:1013]:
```suggestion
  in [Pod topology spread constraints](/docs/concepts/scheduling-eviction/topology-spread-constraints/)
```

Comment (kerthcet) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md]: updategd.

File: content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md (hunk 6 above)
Comment (sftim) [reply content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md:559]: If you'd like to insist this change, at least make it a separate commit (or even separate PR). It's not part of documenting `NodeInclusionPolicyInPodTopologySpread`.

Comment (kerthcet) [reply content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md:559]: hold for https://github.com/kubernetes/website/pull/35966
/hold

**Review by alculquicondor — COMMENTED** [review]
:( I forgot to send my review last week

File: content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md [hunk 9]
```diff
@@ -108,10 +112,13 @@ your cluster. Those fields are:
...
 
-- **topologyKey** is the key of [node labels](#node-labels). If two Nodes are labelled
-  with this key and have identical values for that label, the scheduler treats both
-  Nodes as being in the same topology. The scheduler tries to place a balanced number
-  of Pods into each topology domain.
+- **topologyKey** is the key of [node labels](#node-labels). Nodes that have a label with this key
+	and identical values are considered to be in the same topology.
+	We consider each <key, value> as a "bucket", and try to put balanced number
+	of pods into each bucket.
+	We define a domain as a particular instance of a topology.
```
[the remark this answers is not in the document]
Comment (kerthcet) [reply content/en/docs/concepts/scheduling-eviction/topology-spread-constraints.md:119]: Updated as advised in another PR https://github.com/kubernetes/website/pull/35966, this one only focus on the new feature introduced.

**Comment by alculquicondor** [comment]
I can lgtm after rebase

**Comment by kerthcet** [comment]
Rebased.

**Pull request merged**